test/test_linear_interpolation.py

```diff
@@ -114,3 +114,49 @@ def test_specification_and_derivative():
                         autoderivative = torch.stack(autoderivative).view(*evaluate.shape)
                         assert derivative.shape == autoderivative.shape
                         assert derivative.allclose(autoderivative, atol=1e-5, rtol=1e-5)
+
+
+def test_rectilinear_preparation():
+    # Some fucntions
+    pad_sequence = lambda sequences: torch.nn.utils.rnn.pad_sequence(sequences, batch_first=True, padding_value=nan)
+    ffill = torchcde.misc.torch_ffill
+
+    # Simple test
+    nan = float('nan')
+    t1 = torch.Tensor([0.1, 0.2, 0.9]).view(-1, 1)
+    t2 = torch.Tensor([0.2, 0.3]).view(-1, 1)
+    x1 = torch.Tensor([0.4, nan, 1.1]).view(-1, 1)
+    x2 = torch.Tensor([nan, 2.]).view(-1, 1)
+    x = pad_sequence([torch.cat((t1, x1), -1), torch.cat((t2, x2), -1)])
+    x1_true = torch.Tensor([[0.1, 0.2, 0.2, 0.9, 0.9], [0.4, 0.4, 0.4, 0.4, 1.1]]).T.view(-1, 2)
+    x2_true = torch.Tensor([[0.2, 0.3, 0.3], [nan, nan, 2.]]).T.view(-1, 2)
+    rect_true = ffill(pad_sequence([x1_true, x2_true]))
+    rectilinear = torchcde.interpolation_linear._prepare_rectilinear_interpolation(x, 0)
+    assert rect_true.allclose(rectilinear, equal_nan=True)
+    # Test also if we swap time time dimension
+    x_swap = x[:, :, [1, 0]]
+    rectilinear_swap = torchcde.interpolation_linear._prepare_rectilinear_interpolation(x_swap, 1)
+    assert ffill(rect_true[:, :, [1, 0]]).allclose(rectilinear_swap, equal_nan=True)
+
+    n_channels = 10
+    for _ in range(5):
+        # Build some data with time
+        t_starts = torch.randn(5) ** 2
+        ts = [torch.linspace(t_start, t_start + 10, torch.randint(0, 50, (1,)).item()) for t_start in t_starts]
+        xs = [torch.randn(len(t), n_channels - 1) for t in ts]
+        x = pad_sequence([torch.cat([t_.view(-1, 1), x_], 1) for t_, x_ in zip(ts, xs)])
+        # Add some random nans about the place
+        mask = torch.randint(0, 5, (x.size(0), x.size(1), x.size(2) - 1)).to(float)
+        mask[mask == 0] = float('nan')
+        x[:, :, 1:] = x[:, :, 1:] * mask
+        # Fill
+        x_ffilled = ffill(x)
+        # Compute the true solution
+        N, L, C = x_ffilled.shape
+        rect_true = torch.zeros(N, 2 * L - 1, C)
+        lag = torch.cat([x_ffilled[:, 1:, [0]], x_ffilled[:, :-1, 1:]], -1)
+        rect_true[:, ::2, ] = x_ffilled
+        rect_true[:, 1::2] = lag
+        # Rectilinear solution
+        rectilinear = torchcde.interpolation_linear._prepare_rectilinear_interpolation(x, 0)
+        assert rect_true.allclose(rectilinear, equal_nan=True)
```

Comment:
Also, is it really necessary to call the private function `_prepare...` directly? Can this not be tested via the public interface?

Reply:
Additionally, there should be tests explicitly testing the edge cases of (a) a channel starting with nans; (b) a channel of all nans.

Reply:
I have changed to go via linear_interpolation_coeffs. One thing I noticed though is this will backfill the data if a channel begins with nans. 

It would be good to have the option to fill those nans in with a zero value so as to not pass future information backwards.